sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +35,108 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static int sof_ipc4_build_time_info(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm)
+{
+	struct sof_ipc4_pipeline_registers ppl_reg;
+	struct sof_ipc4_llp_reading_slot llp_slot;
+	struct sof_ipc4_timestamp_info *info;
+	struct sof_ipc4_copier *host_copier = NULL;
+	struct sof_ipc4_copier *dai_copier = NULL;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_dai *dai;
+	u32 node_id;
+	u32 offset;
+	int i;
+
+	/* find host & dai to locate info in memory window */
+	for_each_dapm_widgets(spcm->list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		if (WIDGET_IS_AIF(swidget->widget->id)) {
+			host_copier = swidget->private;
+		} else if (WIDGET_IS_DAI(swidget->widget->id)) {
+			dai = swidget->private;
+			dai_copier = dai->private;
+		}
+	}
+
+	info = spcm->time_info;
+
+	if (!host_copier || !dai_copier) {
+		dev_err(sdev->dev, "failed to find host copier %p or dai copier%p\n",
+			host_copier, dai_copier);
+		return -EINVAL;
+	}
+
+	if (spcm->substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		u64 stream_start_position;
+		u32 dai_sample_size;
+		u32 ch;
+
+		node_id = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
+		if (node_id >= SOF_IPC4_MAX_PIPELINE_REG_SLOTS) {
+			dev_err(sdev->dev, "invalid node_id 0x#%x\n", node_id);
+			return -EINVAL;
+		}
+
+		offset = offsetof(struct sof_ipc4_fw_registers, pipeline_regs) +
+				  node_id * sizeof(ppl_reg);
+		info->ppl_reg_offset = offset;
+		sof_mailbox_read(sdev, sdev->fw_info_box.offset + offset, &ppl_reg,
+				 sizeof(ppl_reg));
+		if (ppl_reg.stream_start_offset != SOF_IPC4_INVALID_STREAM_POSITION) {
+			dev_dbg(sdev->dev, "stream_start_offset position %lld\n",
+				ppl_reg.stream_start_offset);
```

Comment:
Is this log useful? that seems like a very verbose message with limited benefits for debug?

Reply:
it is show the offset return by FW.  I want to check it with cSOF since it is still being developed